File: BaekJoon/Step-by-step troubleshooting(단계별로 문제풀기)/14. 동적계획법1/12865번 평범한 배낭.py
#12865번 평범한 배낭
#import sys
#input = sys.stdin.readline
n,k = map(int,input().split())
wv = [[0]*2 for i in range(n+1)]
start = 98765  
bag = [[0]*(k+1) for i in range(n+1)]

for i in range(1,n+1) :
  wv[i] = list(map(int,input().split()))
  start = min(start,wv[i][0])
  for j in range(start,k+1) :
    if wv[i][0] <= j : 
      bag[i][j] = max(wv[i][1]+bag[i-1][j-wv[i][0]],bag[i-1][j])
    else : 
      bag[i][j] = bag[i-1][j]
print(bag[n][k])

File: BaekJoon/Gold/5/12865번 평범한 배낭.java
//12865번 평범한 배낭.java
package week7;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 평범한_배낭_12865 {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    
    public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine()," ");
		int N = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());
		int wv[][] = new int[N+1][2];
		int start = 10000;
		int[][] bag = new int[N+1][K+1];
		
		for(int i=1;i<N+1;i++) {
			st = new StringTokenizer(br.readLine()," ");
			wv[i][0] = Integer.parseInt(st.nextToken());
			wv[i][1] = Integer.parseInt(st.nextToken());
			start = Math.min(start, wv[i][0]);
			for(int j=start;j<K+1;j++) {
				if(wv[i][0]<=j) {
					bag[i][j]=Math.max(wv[i][1]+bag[i-1][j-wv[i][0]],bag[i-1][j]);
				}
				else {
					bag[i][j]=bag[i-1][j];
				}
			}
		}
		System.out.println(bag[N][K]);
	}
}
